presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -494,27 +508,61 @@ else if (type instanceof MapType) {
+            }
+        }, null);
+
+        // add coercions
+        Expression rewrite = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Expression rewrittenExpression = treeRewriter.defaultRewrite(node, context);
+
+                // cast expression if coercion is registered
+                Type coercion = analysis.getCoercion(node);
+                if (coercion != null) {
+                    rewrittenExpression = new Cast(rewrittenExpression, coercion.getTypeSignature().toString());
+                }
+
+                return rewrittenExpression;
+            }
+        }, expression);
+
         try {
             // expressionInterpreter/optimizer only understands a subset of expression types
             // TODO: remove this when the new expression tree is implemented
-            Expression canonicalized = CanonicalizeExpressions.canonicalizeExpression(expression);
-
-            // verify the expression is constant (has no inputs)
-            ExpressionInterpreter.expressionOptimizer(canonicalized, metadata, session, analysis.getTypes()).optimize(new SymbolResolver() {
-                @Override
-                public Object getValue(Symbol symbol)
-                {
-                    throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain column references");
-                }
-            });
+            Expression canonicalized = CanonicalizeExpressions.canonicalizeExpression(rewrite);
+
+            // The optimization above may have rewritten the expression tree which breaks all the identity maps, so redo the analysis
+            // to re-analyze coercions that might be necessary
+            ExpressionAnalyzer analyzer = ExpressionAnalyzer.createWithoutSubqueries(
+                    metadata.getFunctionRegistry(),
+                    metadata.getTypeManager(),
+                    EXPRESSION_NOT_CONSTANT,
+                    "Constant expression cannot contain as sub-query");
+            analyzer.analyze(canonicalized, new TupleDescriptor(), new AnalysisContext());
 
             // evaluate the expression
-            Object result = ExpressionInterpreter.expressionInterpreter(canonicalized, metadata, session, analysis.getTypes()).evaluate(0);
+            Object result = ExpressionInterpreter.expressionInterpreter(canonicalized, metadata, session, analyzer.getExpressionTypes()).evaluate(0);
             checkState(!(result instanceof Expression), "Expression interpreter returned an unresolved expression");
 
-            return LiteralInterpreter.toExpression(result, analysis.getType(expression));
+            return LiteralInterpreter.toExpression(result, type);
```

Comment:
Instead of passing `type` to this method, return `result` and let callers invoke `LiteralInterpreter.toExpression`